lib/JitEngineDevice.hpp

```diff
@@ -592,31 +641,63 @@ void JitEngineDevice<ImplT>::registerLinkedBinary(
+    ArrayRef<std::unique_ptr<Module>> LinkedModules) {
+  SmallVector<std::string> SHA256HashCodes;
+  for (auto &Mod : LinkedModules) {
+    NamedMDNode *ProteusSHANode =
+        Mod->getNamedMetadata("proteus.module.sha256");
+    assert(ProteusSHANode != nullptr &&
+           "Expected non-null proteus.module.sha256 metadata");
+    assert(ProteusSHANode->getNumOperands() == 1 &&
+           "Hash MD Node should have a single operand");
+    auto MDHash = ProteusSHANode->getOperand(0);
+    MDString *sha256 = dyn_cast<MDString>(MDHash->getOperand(0));
+    if (!sha256) {
+      FATAL_ERROR("Could not read sha256 from module\n");
+    }
+    SHA256HashCodes.push_back(sha256->getString().str());
+    Mod->eraseNamedMetadata(ProteusSHANode);
+  }
+
+  std::sort(SHA256HashCodes.begin(), SHA256HashCodes.end());
+  std::string combinedHash;
+  for (auto hash : SHA256HashCodes) {
+    combinedHash += hash;
+  }
+  return combinedHash;
+}
+
 template <typename ImplT>
 void JitEngineDevice<ImplT>::linkJitModule(
-    Module *M, LLVMContext *Ctx, StringRef KernelName,
-    SmallVector<std::unique_ptr<Module>> &LinkedModules) {
+    Module &M, StringRef KernelName,
+    ArrayRef<std::unique_ptr<Module>> LinkedModules) {
   if (LinkedModules.empty())
     FATAL_ERROR("Expected jit module");
 
-  Linker IRLinker(*M);
-  for (auto &LinkedM : LinkedModules) {
+  Linker IRLinker(M);
+  for (auto &LinkedM : llvm::reverse(LinkedModules)) {
     // Returns true if linking failed.
-    if (IRLinker.linkInModule(std::move(LinkedM), Linker::LinkOnlyNeeded,
-                              [&KernelName](Module &M, const StringSet<> &GVS) {
-                                for (auto &Symbol : GVS) {
-                                  if (Symbol.getKey() == KernelName)
-                                    continue;
-
-                                  Function *F = M.getFunction(Symbol.getKey());
-                                  if (!F)
-                                    continue;
-
-                                  // Internalize functions, the JIT module is
-                                  // self-contained.
-                                  F->setLinkage(GlobalValue::InternalLinkage);
-                                }
-                              }))
+    // We need to clone, as we don't have ownership of the Modules.
+    auto cloned = llvm::CloneModule(*LinkedM);
+
+    if (IRLinker.linkInModule(
+            std::move(cloned), Linker::LinkOnlyNeeded,
+            [&KernelName](Module &Mod, const StringSet<> &GVS) {
+              for (auto &Symbol : GVS) {
+                if (Symbol.getKey() == KernelName)
+                  continue;
+
+                Function *F = Mod.getFunction(Symbol.getKey());
+                if (!F)
+                  continue;
+
+                // Internalize functions, the JIT module is
+                // self-contained.
+                F->setLinkage(GlobalValue::InternalLinkage);
+              }
+            }))
```

Comment:
@ggeorgakoudis I have encountered a weird case on the `linkInModule` functionality. Whether the Linking process will succeed depends on the order of the modules in the `LinkedModules` vector. Is this expected?

Reply:
To verify, I went on the main branch and modified the code:

```
for (auto &LinkedM : LinkedModules) 
```

to 

```
for (auto &LinkedM : llvm::reverse(LinkedModules))
```

and I am getting:
```
ERROR @ /dev/shm/proteus/lib/JitEngineDeviceCUDA.cpp:220 ->  3
Aborted
```

From my investigation, I think the order of Modules in LinkedModules should respect the dependency traversal. In other words, we should not add in the Linker functions before their users are added. Is this something you are aware off? Or did the tests just pass because of random order of registration?

Reply:
Good observation. The IRLinker is stateful and the order matters since we do "link as needed" and internalize symbols through the chain of linking. The current linking is correct because the order in LinkerModuleIds should always have the Module of the kernel-to-JIT first, followed by other device Modules to link in. The error that you see is because the NVPTX compiler failed, most probably due to a missing definition in the final linked Module. I agree this is brittle and we could easily make it order-independent if we just link in everything without internalization, then perform internalization inside the specializeIR stage. Can you open an issue? I'll fix this in a jiff.

Reply:
I think internalizing later is better, there are going to be cases of some cyclic weird dependency and the order of the modules will be hard to describe unless we move the internalization later.